common/src/Microsoft.Azure.IIoT.AspNetCore/src/Auth/Extensions/DataProtectionBuilderEx.cs

```diff
@@ -0,0 +1,218 @@
+
+            // Try reading with app and secret if available.
+            if (!string.IsNullOrEmpty(client.AppId) &&
+                !string.IsNullOrEmpty(client.AppSecret) &&
+                !string.IsNullOrEmpty(client.TenantId)) {
+                keyVault = await TryInititalizeKeyAsync("Application",
+                    vaultUri, keyName, $"RunAs=App; AppId={client.AppId}; " +
+                        $"AppKey={client.AppSecret}; TenantId={client.TenantId}");
+                if (keyVault != null) {
+                    return keyVault;
+                }
+            }
+
+            // Try using aims
+            keyVault = TryInititalizeKeyAsync("Managed Service Identity",
+                vaultUri, keyName).Result;
+            if (keyVault != null) {
+                return keyVault;
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Read configuration secret
+        /// </summary>
+        /// <param name="method"></param>
+        /// <param name="vaultUri"></param>
+        /// <param name="keyName"></param>
+        /// <param name="connectionString"></param>
+        /// <returns></returns>
+        private static async Task<KeyVaultClient> TryInititalizeKeyAsync(
+            string method, string vaultUri, string keyName, string connectionString = null) {
+            var tokenProvider = new AzureServiceTokenProvider(connectionString);
+            try {
+                var keyVaultClient = new KeyVaultClient(
+                    new KeyVaultClient.AuthenticationCallback(
+                        tokenProvider.KeyVaultTokenCallback));
+
+                try {
+                    var key = await keyVaultClient.GetKeyAsync(vaultUri, keyName);
+                }
+                catch {
+                    // Try create key
+                    await keyVaultClient.CreateKeyAsync(vaultUri, keyName, new NewKeyParameters {
+                        KeySize = 2048,
+                        Kty = JsonWebKeyType.Rsa,
+                        KeyOps = new List<string> {
+                            JsonWebKeyOperation.Wrap, JsonWebKeyOperation.Unwrap
+                        }
+                    });
+                }
+                // Worked - we have a working keyvault client.
+                return keyVaultClient;
+            }
+            catch (Exception ex) {
+                Log.Logger.Debug("Failed to authenticate to keyvault {url} using " +
+                    "{method}: {message}",
+                    vaultUri, method, ex.Message);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Dataprotection default configuration
+        /// </summary>
+        internal sealed class DataProtectionConfig : ConfigBase, IClientConfig,
+            IKeyVaultConfig, IStorageConfig {
+
+            /// <summary>Application id</summary>
+            public string AppId => GetStringOrDefault("PCS_KEYVAULT_APPID",
```

Comment:
All three new environment variables (`PCS_KEYVAULT_APPID`, `PCS_KEYVAULT_SECRET` and `PCS_AUTH_TENANT`) should be documented.

Reply:
I added a static class with all currently used PCS_* variables.   That would be the source of the documentation.  I put a work item on the backlog.